src/OmniSharp.LanguageServerProtocol/Handlers/CompletionHandler.cs

```diff
@@ -0,0 +1,144 @@
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using OmniSharp.Extensions.JsonRpc;
+using OmniSharp.Extensions.LanguageServer.Capabilities.Client;
+using OmniSharp.Extensions.LanguageServer.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Mef;
+using OmniSharp.Models.AutoComplete;
+using OmniSharp.Models.TypeLookup;
+
+namespace OmniSharp.LanguageServerProtocol.Handlers
+{
+    [Shared, Export(typeof(CompletionHandler))]
+    class CompletionHandler : ICompletionHandler
+    {
+        public static IEnumerable<IJsonRpcHandler> Enumerate(RequestHandlers handlers)
+        {
+            foreach (var (selector, handler) in handlers
+                .OfType<Mef.IRequestHandler<AutoCompleteRequest, IEnumerable<AutoCompleteResponse>>>())
+                if (handler != null)
+                    yield return new CompletionHandler(handler, selector);
+        }
+
+        private CompletionCapability _capability;
+        private readonly Mef.IRequestHandler<AutoCompleteRequest, IEnumerable<AutoCompleteResponse>> _autoCompleteHandler;
+        private readonly DocumentSelector _documentSelector;
+
+        private static readonly IDictionary<string, CompletionItemKind> _kind = new Dictionary<string, CompletionItemKind>{
+            // types
+            { "Class",  CompletionItemKind.Class },
+            { "Delegate", CompletionItemKind.Class }, // need a better option for this.
+            { "Enum", CompletionItemKind.Enum },
+            { "Interface", CompletionItemKind.Interface },
+            { "Struct", CompletionItemKind.Class }, // TODO: Is struct missing from enum?
+
+            // variables
+            { "Local", CompletionItemKind.Variable },
+            { "Parameter", CompletionItemKind.Variable },
+            { "RangeVariable", CompletionItemKind.Variable },
+
+            // members
+            { "Const", CompletionItemKind.Value }, // TODO: Is const missing from enum?
+            { "EnumMember", CompletionItemKind.Enum },
+            { "Event", CompletionItemKind.Function }, // TODO: Is event missing from enum?
+            { "Field", CompletionItemKind.Field },
+            { "Method", CompletionItemKind.Method },
+            { "Property", CompletionItemKind.Property },
+
+            // other stuff
+            { "Label", CompletionItemKind.Unit }, // need a better option for this.
+            { "Keyword", CompletionItemKind.Keyword },
+            { "Namespace", CompletionItemKind.Module }
+        };
```

Comment:
FWIW, several of the missing items above have been added to VS Code's `CompletionItemKind`, so I expect them to make their way into LSP at some point.

Reply:
Looking at version 4

https://github.com/Microsoft/language-server-protocol/issues/324

Reply:
Yup! That was filed after I asked about LSP's `CompletionItemKind` in https://github.com/Microsoft/vscode/issues/2628.